dnf5-plugins/config-manager_plugin/addrepo.cpp

```diff
@@ -0,0 +1,462 @@
+        const auto urls = libdnf5::OptionStringList(std::vector<std::string>{}).from_string(it->second);
+        if (urls.empty()) {
+            throw ConfigManagerError(M_("Bad baseurl: {}={}"), it->first, it->second);
+        }
+        url = urls[0];
+    }
+    if (url.empty()) {
+        if (auto it = repo_opts.find("mirrorlist"); it != repo_opts.end() && !it->second.empty()) {
+            url = it->second;
+        } else if (auto it = repo_opts.find("metalink"); it != repo_opts.end() && !it->second.empty()) {
+            url = it->second;
+        } else {
+            throw ArgumentParserMissingDependentArgumentError(
+                M_("One of --from-repofile=<URL>, --set=baseurl=<URL>, --set=mirrorlist=<URL>, --set=metalink=<URL> "
+                   "must be set to a non-empty URL"));
+        }
+    }
+
+    if (!repofile_url.empty()) {
+        add_repo_from_repofile(url, repo_dir);
+    } else {
+        create_repo(url, repo_dir);
+    }
+}
+
+
+void ConfigManagerAddRepoCommand::add_repo_from_repofile(
+    const std::string & url, const std::filesystem::path & repo_dir) {
+    auto & ctx = get_context();
+    auto & base = ctx.base;
+    auto logger = base.get_logger();
+
+    logger->info("config-manager: Adding repofile from: {}", url);
+
+    std::string full_url;
+    if (get_url_part(url, CURLUPART_SCHEME) == "") {
+        full_url += "file://" + std::filesystem::absolute(url).string();
+    } else {
+        full_url = url;
+    }
+
+    if (save_filename.empty()) {
+        save_filename = std::filesystem::path(get_url_part(full_url, CURLUPART_PATH)).filename();
+    }
+    if (!save_filename.ends_with(".repo")) {
+        save_filename += ".repo";
+    }
+    auto dest_path = repo_dir / save_filename;
+
+    test_filepath(dest_path);
+
+    auto tmpfilepath = dest_path.string() + ".XXXXXX";
+    auto fd = mkstemp(tmpfilepath.data());
+    if (fd == -1) {
+        throw std::filesystem::filesystem_error(
+            "cannot create temporary file", tmpfilepath, std::error_code(errno, std::system_category()));
+    }
+    close(fd);
+
+    try {
+        try {
+            libdnf5::repo::FileDownloader downloader(base);
+            downloader.add(full_url, tmpfilepath);
+            downloader.download();
+        } catch (const libdnf5::repo::FileDownloadError & e) {
+            throw ConfigManagerError(
+                M_("Failed to download repository configuration file \"{}\": {}"), full_url, std::string{e.what()});
+        }
+
+        libdnf5::ConfigParser parser;
+        parser.read(tmpfilepath);
+        std::vector<std::string> repo_ids;
+        repo_ids.reserve(parser.get_data().size());
+        for (const auto & [repo_id, opts] : parser.get_data()) {
+            repo_ids.emplace_back(repo_id);
+        }
+        test_if_ids_not_already_exist(repo_ids, dest_path);
+
+        // Test if the repository options can be set.
```

Comment:
I like the test. There might be some legacy configuration option that might cause the problem. As I said, this is a good approach.